File: app/src/main/java/com/azizbekian/spyur/converter/ListingConverter.java
package com.azizbekian.spyur.converter;

import android.content.Context;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.lang.annotation.Annotation;
import java.lang.reflect.Type;

import com.azizbekian.spyur.SpyurApplication;
import com.azizbekian.spyur.model.LegendEnum;
import com.azizbekian.spyur.model.ListingResponse;
import com.azizbekian.spyur.model.MhdEnum;
import com.azizbekian.spyur.model.WeekDayEnum;
import com.azizbekian.spyur.rest.SpyurApi;
import com.azizbekian.spyur.utils.LogUtils;
import okhttp3.ResponseBody;
import retrofit2.Converter;
import retrofit2.Retrofit;

import static com.azizbekian.spyur.model.ListingResponse.WorkingDays.Day.FRI;
import static com.azizbekian.spyur.model.ListingResponse.WorkingDays.Day.MON;
import static com.azizbekian.spyur.model.ListingResponse.WorkingDays.Day.SAT;
import static com.azizbekian.spyur.model.ListingResponse.WorkingDays.Day.SUN;
import static com.azizbekian.spyur.model.ListingResponse.WorkingDays.Day.THU;
import static com.azizbekian.spyur.model.ListingResponse.WorkingDays.Day.TUE;
import static com.azizbekian.spyur.model.ListingResponse.WorkingDays.Day.WED;

/**
 * Created on May 09, 2016.
 *
 * @author Andranik Azizbekian (andranik.azizbekyan@gmail.com)
 */
public class ListingConverter implements Converter<ResponseBody, ListingResponse> {

    private static final String PREFIX_PHONE_NUMBER = "•";

    private Context mContext;

    public static final class Factory extends Converter.Factory {

        @Override
        public Converter<ResponseBody, ?> responseBodyConverter(Type type, Annotation[] annotations,
                                                                Retrofit retrofit) {
            return INSTANCE;
        }

    }

    private ListingConverter() {
        // mContext = SpyurApplication.getComponent().getApplicationContext();
    }

    static final ListingConverter INSTANCE = new ListingConverter();

    @Override
    public ListingResponse convert(ResponseBody value) throws IOException {

        ListingResponse.Builder builder = new ListingResponse.Builder();

        Document document = Jsoup.parse(value.string(), SpyurApi.ENDPOINT);
        final Elements fieldSets = document.select("fieldset");
        if (null == fieldSets) return null;

        for (Element fieldSet : fieldSets) {
            LegendEnum legendEnum = LegendEnum.fromString(mContext, fieldSet.select("legend").text());
            if (null == legendEnum) continue;

            switch (legendEnum) {
                case CONTACT_INFORMATION:
                    parseContactInformation(fieldSet, builder);
                    break;
            }
        }

        return builder.build();
    }

    private void parseContactInformation(Element fieldSet, ListingResponse.Builder builder) {
        final Element ulClassContactInfo = fieldSet.select("ul[class=contactInfo]").first();
        if (null != ulClassContactInfo) {
            final Elements mhds = ulClassContactInfo.select("li[class=mhd]");
            if (null == mhds) return;
            for (Element e : mhds) {
                MhdEnum mhdEnum = MhdEnum.fromString(mContext, e.text());
                if (null == mhdEnum) continue;
                switch (mhdEnum) {
                    case EXECUTIVE:
                        parseExecutives(e.nextElementSibling(), builder);
                        break;
                    case ADDRESS_TELEPHONE:
                        parseAddressTelephone(e.nextElementSibling(), builder);
                        break;
                    case WORKING_DAYS_HOURS:
                        parseWorkingDays(e.nextElementSibling(), builder);
                        break;
                    case WEBSITE:
                        parseWebsite(e.nextElementSibling(), builder);
                        break;
                    case LISTING_IN_SPYUR:
                        parseListingInSpyur(e.nextElementSibling(), builder);
                        break;
                }
            }

            final Elements images = fieldSet.select("div[class=div-img]");
            if (!images.isEmpty()) {
                parseImages(images, builder);
            }

            final Elements videos = fieldSet.select("div#Slide_video");
            if (!videos.isEmpty()) {
                parseVideos(videos, builder);
            }
        }
    }

    private void parseVideos(Elements videos, ListingResponse.Builder builder) {
        builder.setVideoUrl(videos.select("a[class=fl curimg]").first().attr("id"));
    }

    private void parseImages(Elements images, ListingResponse.Builder builder) {
        for (Element imageDiv : images) {
            builder.addImage(imageDiv.select("a").first().attr("abs:href"));
        }
    }

    private void parseListingInSpyur(Element element, ListingResponse.Builder builder) {
        if (null == element) return;
        builder.setListingInSpyur(element.text());
    }

    private void parseWebsite(Element element, ListingResponse.Builder builder) {
        if (null == element) return;

        Elements websiteDivs = element.select("div > div");
        if (null == websiteDivs) return;

        for (Element div : websiteDivs) {
            builder.addWebsite(div.text());
        }
    }

    private void parseWorkingDays(Element element, ListingResponse.Builder builder) {
        if (null == element) return;

        Element tr = element.select("tr").first();
        if (null == tr) return;

        // removing &nbsp;
        String workingDays = tr.select("td").get(0).text().replace("\u00a0", "").trim();
        String[] splittedWorkDays = workingDays.split("\\s+");

        ListingResponse.WorkingDays.Builder workingDaysBuilder
                = new ListingResponse.WorkingDays.Builder();

        for (String s : splittedWorkDays) {
            WeekDayEnum weekDayEnum = WeekDayEnum.fromString(mContext, s);
            if (null == weekDayEnum) continue;
            switch (weekDayEnum) {
                case MON:
                    workingDaysBuilder.setWorkingDay(MON);
                    break;
                case TUE:
                    workingDaysBuilder.setWorkingDay(TUE);
                    break;
                case WED:
                    workingDaysBuilder.setWorkingDay(WED);
                    break;
                case THU:
                    workingDaysBuilder.setWorkingDay(THU);
                    break;
                case FRI:
                    workingDaysBuilder.setWorkingDay(FRI);
                    break;
                case SAT:
                    workingDaysBuilder.setWorkingDay(SAT);
                    break;
                case SUN:
                    workingDaysBuilder.setWorkingDay(SUN);
                    break;
            }
        }

        builder.setWorkingDays(workingDaysBuilder.build());
    }

    private void parseAddressTelephone(Element address, ListingResponse.Builder builder) {
        if (null == address) return;

        Elements divs = address.select("li > div");
        if (null == divs) return;
        ListingResponse.ContactInfo.Builder contactInfoBuilder;
        for (Element div : divs) {
            contactInfoBuilder = new ListingResponse.ContactInfo.Builder();
            Element a = div.select("div > a").first();
            if (null != a) {
                try {
                    double lat = Double.parseDouble(a.attr("lat"));
                    double lon = Double.parseDouble(a.attr("lon"));
                    contactInfoBuilder.setLocation(lat, lon);
                } catch (NumberFormatException ignored) {
                }
            }
            contactInfoBuilder.setAddress(div.ownText());

            // Region information
            Elements childDivs = div.select("div > div");
            if (null != childDivs && childDivs.size() > 0) {

                Element e = childDivs.first();
                do {
                    String text = e.text();
                    if (!text.startsWith(PREFIX_PHONE_NUMBER)) {
                        contactInfoBuilder.setRegion(text.substring(1, text.length() - 1));
                    } else {
                        try {
                            contactInfoBuilder.addPhoneNumber(text
                                    .substring(1)
                                    .trim()
                                    .replace("-", ""));
                        } catch (IndexOutOfBoundsException ignored) {
                            LogUtils.e(ignored.getMessage());
                        }
                    }
                    e = e.nextElementSibling();
                } while (e != null);

            }
            builder.addContactInfo(contactInfoBuilder.build());
        }
    }

    private void parseExecutives(Element executives, ListingResponse.Builder builder) {
        if (null == executives) return;

        Element li = executives.select("li").first();
        if (null != li) {
            Elements divs = li.select("div");
            if (null == divs) return;
            for (Element e : divs) {
                builder.addExecutive(e.text());
            }
        }
    }

}

File: app/src/main/kotlin/com/incentive/yellowpages/misc/converter/ListingConverter.kt
package com.incentive.yellowpages.misc.converter

import android.content.Context
import com.incentive.yellowpages.data.model.LegendEnum
import com.incentive.yellowpages.data.model.ListingResponse
import com.incentive.yellowpages.data.model.MhdEnum
import com.incentive.yellowpages.data.model.WeekDayEnum
import com.incentive.yellowpages.data.remote.ApiContract
import com.incentive.yellowpages.injection.ApplicationContext
import com.incentive.yellowpages.ui.base.BaseApplication
import okhttp3.ResponseBody
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import org.jsoup.select.Elements
import retrofit2.Converter
import retrofit2.Retrofit
import timber.log.Timber
import java.io.IOException
import java.lang.reflect.Type

class ListingConverter : Converter<ResponseBody, ListingResponse> {

    companion object {

        private val PREFIX_PHONE_NUMBER = "•"
        private val PREFIX_DISTRICT = "("

        internal val INSTANCE = ListingConverter()
    }

    @ApplicationContext val context: Context = BaseApplication.appComponent.context()

    class Factory : Converter.Factory() {

        override fun responseBodyConverter(type: Type?, annotations: Array<Annotation>?,
                                           retrofit: Retrofit?): Converter<ResponseBody, *> {
            return INSTANCE
        }

    }

    @Throws(IOException::class)
    override fun convert(value: ResponseBody): ListingResponse? {

        val builder = ListingResponse.Builder()

        val document = Jsoup.parse(value.string(), ApiContract.BASE)
        val fieldSets = document.select("fieldset") ?: return null

        for (fieldSet in fieldSets) {
            val legendEnum = LegendEnum.fromString(context, fieldSet.select("legend").text()) ?: continue

            when (legendEnum) {
                LegendEnum.CONTACT_INFORMATION -> parseContactInformation(fieldSet, builder)
            }
        }

        return builder.build()
    }

    private fun parseContactInformation(fieldSet: Element, builder: ListingResponse.Builder) {
        val ulClassContactInfo = fieldSet.select("ul[class=contactInfo]").first()
        if (null != ulClassContactInfo) {
            val mhds = ulClassContactInfo.select("li[class=mhd]") ?: return
            for (e in mhds) {
                val mhdEnum = MhdEnum.fromString(context, e.text()) ?: continue
                when (mhdEnum) {
                    MhdEnum.EXECUTIVE -> parseExecutives(e.nextElementSibling(), builder)
                    MhdEnum.ADDRESS_TELEPHONE -> parseAddressTelephone(e.nextElementSibling(), builder)
                    MhdEnum.WEBSITE -> parseWebsite(e.nextElementSibling(), builder)
                    MhdEnum.LISTING_IN_SPYUR -> parseListingInSpyur(e.nextElementSibling(), builder)
                }
            }

            val images = fieldSet.select("div[class=div-img]")
            if (!images.isEmpty()) {
                parseImages(images, builder)
            }

            val videos = fieldSet.select("div#Slide_video")
            if (!videos.isEmpty()) {
                parseVideos(videos, builder)
            }
        }
    }

    private fun parseVideos(videos: Elements, builder: ListingResponse.Builder) {
        builder.setVideoUrl(videos.select("a[class=fl curimg]").first().attr("id"))
    }

    private fun parseImages(images: Elements, builder: ListingResponse.Builder) {
        for (imageDiv in images) {
            builder.addImage(imageDiv.select("a").first().attr("abs:href"))
        }
    }

    private fun parseListingInSpyur(element: Element?, builder: ListingResponse.Builder) {
        if (null == element) return
        builder.setListingInSpyur(element.text())
    }

    private fun parseWebsite(element: Element?, builder: ListingResponse.Builder) {
        if (null == element) return

        val websiteDivs = element.select("div > div") ?: return

        for (div in websiteDivs) {
            builder.addWebsite(div.text())
        }
    }

    private fun parseAddressTelephone(address: Element?, builder: ListingResponse.Builder) {
        if (null == address) return

        val divs = address.select("li > div") ?: return
        var contactInfoBuilder: ListingResponse.ContactInfo.Builder
        for (div in divs) {
            contactInfoBuilder = ListingResponse.ContactInfo.Builder()
            val a = div.select("div > a").first()
            if (null != a) {
                try {
                    val lat = java.lang.Double.parseDouble(a.attr("lat"))
                    val lon = java.lang.Double.parseDouble(a.attr("lon"))
                    contactInfoBuilder.setLocation(lat, lon)
                } catch (ignored: NumberFormatException) {
                }

            }
            contactInfoBuilder.setAddress(div.ownText())

            // Region information
            val childDivs = div.select("div > div")
            if (null != childDivs && childDivs.size > 0) {

                var e: Element? = childDivs.first()
                do {
                    val text = e!!.text()
                    if (text.startsWith(PREFIX_DISTRICT)) {
                        contactInfoBuilder.setRegion(text.substring(1, text.length - 1))
                    } else if (text.startsWith(PREFIX_PHONE_NUMBER)) {
                        try {
                            contactInfoBuilder.addPhoneNumber(text
                                    .substring(1)
                                    .trim { it <= ' ' }
                                    .replace("-", ""))
                        } catch (ignored: IndexOutOfBoundsException) {
                            Timber.e(ignored.message)
                        }
                    } else {
                        parseWorkingDays(e, builder)
                    }
                    e = e.nextElementSibling()
                } while (e != null)

            }
            builder.addContactInfo(contactInfoBuilder.build())
        }
    }

    private fun parseWorkingDays(element: Element?, builder: ListingResponse.Builder) {
        if (null == element) return

        val splittedWorkDays = element.text().split("\\s+".toRegex())

        val workingDaysBuilder = ListingResponse.WorkingDays.Builder()

        for (workingDay in splittedWorkDays) {
            if (workingDay.length > 3) continue
            val weekDayEnum = WeekDayEnum.fromString(context, workingDay)

            when (weekDayEnum) {
                WeekDayEnum.MON -> workingDaysBuilder.setWorkingDay(ListingResponse.WorkingDays.Day.MON)
                WeekDayEnum.TUE -> workingDaysBuilder.setWorkingDay(ListingResponse.WorkingDays.Day.TUE)
                WeekDayEnum.WED -> workingDaysBuilder.setWorkingDay(ListingResponse.WorkingDays.Day.WED)
                WeekDayEnum.THU -> workingDaysBuilder.setWorkingDay(ListingResponse.WorkingDays.Day.THU)
                WeekDayEnum.FRI -> workingDaysBuilder.setWorkingDay(ListingResponse.WorkingDays.Day.FRI)
                WeekDayEnum.SAT -> workingDaysBuilder.setWorkingDay(ListingResponse.WorkingDays.Day.SAT)
                WeekDayEnum.SUN -> workingDaysBuilder.setWorkingDay(ListingResponse.WorkingDays.Day.SUN)
            }
        }

        builder.setWorkingDays(workingDaysBuilder.build())
    }

    private fun parseExecutives(executives: Element?, builder: ListingResponse.Builder) {
        if (null == executives) return

        val li = executives.select("li").first()
        if (null != li) {
            val divs = li.select("div") ?: return
            for (e in divs) {
                builder.addExecutive(e.text())
            }
        }
    }

}
